cpu/fe310/periph/pwm.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2019 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_fe310
+ * @ingroup     drivers_periph_pwm
+ * @{
+ *
+ * @file
+ * @brief       PWM driver implementation for the FE310
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ * @}
+ */
+
+#include <assert.h>
+
+#include "cpu.h"
+#include "periph/pwm.h"
+#include "periph/gpio.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define PWM_CMPSIZE     (0x4)
+#define PWM_CMPWIDTH    (16)
+
+uint32_t pwm_init(pwm_t dev, pwm_mode_t mode, uint32_t freq, uint16_t res)
+{
+    assert(dev < PWM_NUMOF);
+
+    /* Compute PWM counter scale */
+    uint8_t scale = 0;
+    while ((scale < (PWM_CMPWIDTH - 1)) &&
+           (freq * res) > (uint32_t)(1 << (scale + PWM_CMPWIDTH))) {
+        scale++;
+    }
```

Comment:
This calculation seems very bad to me, how can you get a good approximation of `freq` without taking the `cpu_freq` into account? It results in very weird frequency settings that do not make much sense.

If you want full resolution for example that means you can only scale the register. But if you wan't only lets say 8 bits of resolution then you have 8 more to tune the desired frequency. For me the calculation should take into account:

- bits left over by desired resolution: `res/2^16`
- `cpu_freq`

 We are basically trying to resolve: 

`min CPU_FREQ * (2^x)/(2^(16-scale)) - freq`, where `x = 1...2^rembits` and `scale = 0...(PWM_CMPWIDTH - 1)`. I don't know efficient algorithms to find this, would have to think about it more. But one way would be find a scale which gives you a `PWM_FREQ` higher than `freq` and then trim the value with the remaining resolution bits. Another option is for every scale find the best match and then compare them all.

For the first suggestion.


1. In pseudo code: find scaling that fives you a `PWM_FREQ > freq`:

```
    uint8_t scale = (PWM_CMPWIDTH - 1);
    while ((scale > 0) &&
           cpu_freq() / (uint32_t)(1 << (scale + PWM_CMPWIDTH)) << freq) {
        scale--;
    }
```

2.  Use the leftover bits for the required resolution to trim `PWM_FREQ` as close to `freq` as possible

Then those leftover bits or rem_res give you more wiggle room to set the a freq closest to the desired one.